### PR TITLE
Php update

### PR DESCRIPTION
**Pull request by rvwatch**
Updates workbench to php 7.2 via composer.json "php": "^7.2". Also updates to the latest version of Xdebug which should now function properly in VS code. Manual testing on the staging environment to follow.

### CODE REVIEW

**Comment by vazexqi** [comment]
@rvwatch - Worth reading https://chris.beams.io/posts/git-commit/ on how to write good commit messages.

File: composer.lock [hunk 1]
```diff
@@ -4,20 +4,20 @@
         "Read more about it at https://getcomposer.org/doc/01-basic-usage.md#installing-dependencies",
         "This file is @generated automatically"
     ],
-    "content-hash": "68ec110aab4b7d444c66d090d7252e4a",
+    "content-hash": "8c1127e8c3f580a05e127400b65efd46",
```
Comment (vazexqi) [inline composer.lock]: How did you generate this in the end? We should document/automate this so that it's easy for someone else to do this.

File: README.md [hunk 2]
```diff
@@ -12,6 +12,9 @@ such as customizable SOAP headers, debug logs for API traffic, backward
 compatibility testing with previous API versions, and single sign-on
 integration within the Salesforce application.
 
+
```
Comment (vazexqi) [inline README.md]: Let's remove these lines.

File: workbench/config/WorkbenchConfig.php [hunk 3]
```diff
@@ -52,33 +52,35 @@ function __construct() {
         // load environment variable based overrides
         $configNamespace = "forceworkbench";
         $configDelim = "__";
+
```
Comment (vazexqi) [inline workbench/config/WorkbenchConfig.php]: Could you clean up this file so that the diff looks cleaner? There are too many whitespace changes.

File: composer.json [hunk 4]
```diff
@@ -1,6 +1,6 @@
 {
     "require": {
-        "php": "~5.6.12",
+        "php": "^7.2",
```
Comment (DatGuyJonathan) [inline composer.json]: We should probably use `~` so we don't get accidentally upgraded to the latest 7.x.

Comment (rvwatch) [reply composer.json]: Ok. I'll set it back to ~.

File: composer.lock (hunk 1 above)
Comment (rvwatch) [reply composer.lock]: I followed along this link: https://devcenter.heroku.com/articles/deploying-php
Ended up going with composer update. I'm not sure how to automate that just yet. It appears as though the users get this lock file when they clone the project. If that's the case why would we need to automate this? Obviously, I'm not sure what "content-hash" is up to, so I'm happy to dig through this some more.

File: workbench/config/WorkbenchConfig.php [hunk 5]
```diff
@@ -68,7 +68,7 @@ function __construct() {
                 }
         
                 if (!isset($point[$keyPart])) {
-                    $point[$keyPart] = "";
+                    $point[$keyPart] = [];
```
Comment (vazexqi) [inline workbench/config/WorkbenchConfig.php]:
```suggestion
                    $point[$keyPart] = "";
```

File: README.md (hunk 2 above)
Comment (vazexqi) [reply README.md]: @rvwatch ^

File: composer.lock (hunk 1 above)
Comment (vazexqi) [reply composer.lock]: The part that I want to automate is the composer.lock generation so that you, me, Jonny, etc can generate it consistently.

Right now, having to do things locally is not consistent. We need a consistent way to do this just like how we use docker to standardize our local development.

I would like us to use https://hub.docker.com/r/library/composer/. I've done it and I can show you how to do this.

File: composer.lock [hunk 6]
```diff
@@ -78,7 +78,7 @@
     "prefer-stable": false,
     "prefer-lowest": false,
     "platform": {
-        "php": "~5.6.12",
+        "php": "^7.2",
```
Comment (vazexqi) [inline composer.lock]: Case in point. You changed composer.json but forgot to change this one.

File: workbench/config/WorkbenchConfig.php [hunk 7]
```diff
@@ -53,6 +53,8 @@ function __construct() {
         $configNamespace = "forceworkbench";
         $configDelim = "__";
         foreach ($_ENV as $envKey => $envValue) {
+
```
Comment (vazexqi) [inline workbench/config/WorkbenchConfig.php]: @rvwatch - Could we minimize unnecessary whitespace changes to the file? Thsi makes the diff very hard to follow.

File: workbench/config/overrides.php [hunk 8]
```diff
@@ -43,5 +43,4 @@
 //
 // $config["orgIdWhiteList"]["default"] = "00D000000000001, 00D000000000002";
 // $config["orgIdBlackList"]["default"] = "00D000000000003";
-
```
Comment (vazexqi) [inline workbench/config/overrides.php]: Please revert this file since there are no functional changes here.

File: workbench/config/WorkbenchConfig.php (hunk 7 above)
Comment (rvwatch) [reply workbench/config/WorkbenchConfig.php]: Is there a linter or approved php formatter that I can use so that I can keep this from happening? You gotta believe me when I say I'm not trying to add these... lol

**Comment by rvwatch** [comment]
I see the whitespace issues. I thought I had reverted properly. I'll get try to get those setup correctly.

File: workbench/config/WorkbenchConfig.php [hunk 9]
```diff
@@ -60,17 +60,23 @@ function __construct() {
             $envKey = str_replace("___DOT___", ".", $envKey);
         
             $envKeyParts = explode($configDelim, $envKey);
-        
+
+            $lastKey = end($envKeyParts);
+            reset($envKeyParts);
```
Comment (vazexqi) [inline workbench/config/WorkbenchConfig.php]: @rvwatch: Could you explain why we use this way to get the end of the array (get the end then reset the array)? Is this idiomatic PHP?

File: scripts/composer_update.sh [hunk 10]
```diff
@@ -0,0 +1 @@
+docker run --rm --interactive --tty --volume $PWD:/app composer update --ignore-platform-reqs --no-scripts
```
Comment (vazexqi) [inline scripts/composer_update.sh]: Should add the [shebang](https://en.wikipedia.org/wiki/Shebang_(Unix)) line to this.
Also, did you remember to `chmod +x` before committing it? I can't tell from the web interface.

File: workbench/config/WorkbenchConfig.php (hunk 9 above)
Comment (rvwatch) [reply workbench/config/WorkbenchConfig.php]: I've been reading some more on this. Though end()/reset() is a relatively common pattern, I don't like the idea of messing with the internal pointer of the original array and having to reset it. That feels flimsy. I think switching it to something like: `end((array_values($array)))` may be a better solution as it doesn't touch the pointer of the original array. No need to reset and it should be fine performance wise as it only hits small arrays. 

I also found this interesting stack overflow comment if you're curious. https://stackoverflow.com/a/41795859

File: workbench/config/WorkbenchConfig.php [hunk 11]
```diff
@@ -60,17 +60,22 @@ function __construct() {
             $envKey = str_replace("___DOT___", ".", $envKey);
         
             $envKeyParts = explode($configDelim, $envKey);
-        
+
+            $lastKey = end(array_values($envKeyParts));
```
Comment (vazexqi) [inline workbench/config/WorkbenchConfig.php]: @rvwatch - Actually let's **not** do this. The old way, while a bit clunky seems to be fastest. According to https://theghostwhocodes.com/php-how-to-get-the-last-element-of-an-array-2a0db5343520

> end(array_values($array)); It’ s the worst solution, both in terms of elegance and performance: it takes 176.85 s! Fast using end to work on array internal pointer, but terribly slow creating a new array every time with array_values, used only to avoid the reset call.

Comment (rvwatch) [reply workbench/config/WorkbenchConfig.php]: Ok. I'll switch it back up.

**Comment by jonnyhork** [comment]
I chose not to delete the `phpUpdate` branch after the merge until the testing is done.
